### PR TITLE
HDDS-1422. Exception during DataNode shutdown. Contributed by Arpit A…

### PR DESCRIPTION
**Pull request by arp7**
…garwal.

Change-Id: I981fbd087baca80cc6b4ff58e91e63dcd9726c13

### CODE REVIEW

**Comment by bharatviswa504** [comment]
Hi @arp7 
Thanks for the fix.
To understand the fix, we have removed the setting of usage to null, so that even when DN is shut down, and if someone calls volume usage we return old value, instead of throwing an exception.

Is my understanding correct or am I missing something here?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 25 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1058 | trunk passed |
| +1 | compile | 38 | trunk passed |
| +1 | checkstyle | 20 | trunk passed |
| +1 | mvnsite | 33 | trunk passed |
| +1 | shadedclient | 711 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 51 | trunk passed |
| +1 | javadoc | 28 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 36 | the patch passed |
| +1 | compile | 26 | the patch passed |
| +1 | javac | 26 | the patch passed |
| -0 | checkstyle | 13 | hadoop-hdds/container-service: The patch generated 3 new + 0 unchanged - 0 fixed = 3 total (was 0) |
| +1 | mvnsite | 30 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 729 | patch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 57 | hadoop-hdds/container-service generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | javadoc | 25 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 77 | container-service in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3072 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-hdds/container-service |
|  |  Inconsistent synchronization of org.apache.hadoop.ozone.container.common.volume.VolumeUsage.scmUsage; locked 60% of time  Unsynchronized access at VolumeUsage.java:60% of time  Unsynchronized access at VolumeUsage.java:[line 183] |
| Failed junit tests | hadoop.ozone.container.common.TestDatanodeStateMachine |
|   | hadoop.ozone.container.common.volume.TestVolumeSet |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/725 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 21d98372751e 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 586826f |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/artifact/out/diff-checkstyle-hadoop-hdds_container-service.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/artifact/out/new-findbugs-hadoop-hdds_container-service.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/artifact/out/patch-unit-hadoop-hdds_container-service.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/testReport/ |
| Max. process+thread count | 432 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by arp7** [comment]
> Hi @arp7
> Thanks for the fix.
> To understand the fix, we have removed the setting of usage to null, so that even when DN is shut down, and if someone calls volume usage we return old value, instead of throwing an exception.
> 
> Is my understanding correct or am I missing something here?

That is correct @bharatviswa504 .

**Comment by bharatviswa504** [comment]
Thank You @arp7 for the reply.
Now I want to understand what benefit, this fix will give. If I understand correctly, if during shutdown even if some one is using volume API to get usage they will get some value, instead of exception. Are is there any other thing?

**Comment by arp7** [comment]
This is to avoid spurious exception messages during DataNode shutdown. They were creating noise in the logs, especially when running MiniOzoneChaosCluster which frequently restarts nodes.

**Comment by arp7** [comment]
Addressed issues flagged by CI, and rebased to current trunk.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 51 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 435 | trunk passed |
| +1 | compile | 212 | trunk passed |
| +1 | checkstyle | 52 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 841 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 131 | trunk passed |
| 0 | spotbugs | 325 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 567 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 464 | the patch passed |
| +1 | compile | 251 | the patch passed |
| +1 | javac | 251 | the patch passed |
| +1 | checkstyle | 65 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 769 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 155 | the patch passed |
| +1 | findbugs | 558 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 155 | hadoop-hdds in the patch failed. |
| -1 | unit | 1344 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 39 | The patch does not generate ASF License warnings. |
| | | 6961 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.TestHddsDatanodeService |
|   | hadoop.ozone.container.common.volume.TestVolumeSet |
|   | hadoop.ozone.container.common.volume.TestVolumeSetDiskChecks |
|   | hadoop.ozone.container.keyvalue.TestKeyValueContainer |
|   | hadoop.ozone.container.common.TestDatanodeStateMachine |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestCloseContainerCommandHandler |
|   | hadoop.ozone.container.keyvalue.TestBlockManagerImpl |
|   | hadoop.ozone.container.keyvalue.TestKeyValueBlockIterator |
|   | hadoop.ozone.container.ozoneimpl.TestOzoneContainer |
|   | hadoop.ozone.container.common.volume.TestHddsVolume |
|   | hadoop.ozone.container.keyvalue.TestKeyValueContainerMarkUnhealthy |
|   | hadoop.ozone.container.common.volume.TestRoundRobinVolumeChoosingPolicy |
|   | hadoop.ozone.container.common.impl.TestHddsDispatcher |
|   | hadoop.ozone.container.keyvalue.TestKeyValueContainerCheck |
|   | hadoop.ozone.container.keyvalue.TestKeyValueHandler |
|   | hadoop.ozone.container.keyvalue.TestChunkManagerImpl |
|   | hadoop.ozone.client.rpc.TestBlockOutputStream |
|   | hadoop.ozone.scm.pipeline.TestPipelineManagerMXBean |
|   | hadoop.ozone.web.TestOzoneWebAccess |
|   | hadoop.ozone.client.rpc.TestBCSID |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.scm.node.TestQueryNode |
|   | hadoop.ozone.web.client.TestOzoneClient |
|   | hadoop.ozone.om.TestOmInit |
|   | hadoop.ozone.TestStorageContainerManager |
|   | hadoop.hdds.scm.pipeline.TestSCMPipelineManager |
|   | hadoop.ozone.client.rpc.TestHybridPipelineOnDatanode |
|   | hadoop.ozone.container.common.TestBlockDeletingService |
|   | hadoop.ozone.container.TestContainerReplication |
|   | hadoop.ozone.TestContainerStateMachineIdempotency |
|   | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion |
|   | hadoop.ozone.container.metrics.TestContainerMetrics |
|   | hadoop.ozone.client.rpc.TestContainerStateMachine |
|   | hadoop.ozone.scm.node.TestSCMNodeMetrics |
|   | hadoop.ozone.om.TestOzoneManager |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.om.TestContainerReportWithKeys |
|   | hadoop.ozone.om.TestMultipleContainerReadWrite |
|   | hadoop.ozone.container.common.impl.TestContainerPersistence |
|   | hadoop.hdds.scm.container.TestContainerStateManagerIntegration |
|   | hadoop.hdds.scm.pipeline.TestNode2PipelineMap |
|   | hadoop.ozone.TestContainerOperations |
|   | hadoop.ozone.container.ozoneimpl.TestOzoneContainer |
|   | hadoop.ozone.scm.TestXceiverClientMetrics |
|   | hadoop.ozone.scm.TestAllocateContainer |
|   | hadoop.ozone.container.ozoneimpl.TestSecureOzoneContainer |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis |
|   | hadoop.ozone.web.client.TestKeys |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestCloseContainerHandler |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestDeleteContainerHandler |
|   | hadoop.ozone.scm.pipeline.TestSCMPipelineMetrics |
|   | hadoop.ozone.ozShell.TestOzoneDatanodeShell |
|   | hadoop.ozone.scm.TestGetCommittedBlockLengthAndPutKey |
|   | hadoop.ozone.om.TestOmBlockVersioning |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineUtils |
|   | hadoop.ozone.om.TestOmAcls |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestCloseContainerByPipeline |
|   | hadoop.ozone.om.TestOmMetrics |
|   | hadoop.ozone.client.rpc.TestCloseContainerHandlingByClient |
|   | hadoop.ozone.client.rpc.TestContainerStateMachineFailures |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.om.TestOMDbCheckpointServlet |
|   | hadoop.ozone.ozShell.TestS3Shell |
|   | hadoop.ozone.scm.TestContainerSmallFile |
|   | hadoop.ozone.scm.TestXceiverClientManager |
|   | hadoop.hdds.scm.pipeline.TestPipelineClose |
|   | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.client.rpc.TestReadRetries |
|   | hadoop.ozone.scm.TestSCMNodeManagerMXBean |
|   | hadoop.ozone.web.TestOzoneRestWithMiniCluster |
|   | hadoop.ozone.scm.TestSCMMXBean |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.TestOzoneClientRetriesOnException |
|   | hadoop.hdds.scm.safemode.TestSCMSafeModeWithPipelineRules |
|   | hadoop.ozone.ozShell.TestOzoneShell |
|   | hadoop.ozone.om.TestOzoneManagerRestInterface |
|   | hadoop.hdds.scm.pipeline.TestNodeFailure |
|   | hadoop.ozone.web.TestOzoneVolumes |
|   | hadoop.ozone.web.client.TestBuckets |
|   | hadoop.ozone.web.client.TestVolume |
|   | hadoop.ozone.client.rpc.TestCommitWatcher |
|   | hadoop.hdds.scm.pipeline.TestSCMRestart |
|   | hadoop.ozone.web.client.TestKeysRatis |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-725/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/725 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 4937e4c93966 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3418bbb |
| Default Java | 1.8.0_191 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/2/testReport/ |
| Max. process+thread count | 5010 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-725/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by arp7** [comment]
Fix UT failures caused by previous commit.

File: phantomjsdriver.log [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[INFO  - 2019-05-13T20:32:17.166Z] GhostDriver - Main - running on port 3369
+[INFO  - 2019-05-13T20:32:17.424Z] Session [332af4c0-75be-11e9-ae47-adcdd47c587b] - page.settings - {"XSSAuditingEnabled":false,"javascriptCanCloseWindows":true,"javascriptCanOpenWindows":true,"javascriptEnabled":true,"loadImages":true,"localToRemoteUrlAccessEnabled":false,"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/538.1 (KHTML, like Gecko) PhantomJS/2.1.1 Safari/538.1","webSecurityEnabled":true}
+[INFO  - 2019-05-13T20:32:17.424Z] Session [332af4c0-75be-11e9-ae47-adcdd47c587b] - page.customHeaders:  - {}
+[INFO  - 2019-05-13T20:32:17.425Z] Session [332af4c0-75be-11e9-ae47-adcdd47c587b] - Session.negotiatedCapabilities - {"browserName":"phantomjs","version":"2.1.1","driverName":"ghostdriver","driverVersion":"1.2.0","platform":"mac-unknown-64bit","javascriptEnabled":true,"takesScreenshot":true,"handlesAlerts":false,"databaseEnabled":false,"locationContextEnabled":false,"applicationCacheEnabled":false,"browserConnectionEnabled":false,"cssSelectorsEnabled":true,"webStorageEnabled":false,"rotatable":false,"acceptSslCerts":false,"nativeEvents":true,"proxy":{"proxyType":"direct"}}
```
Comment (bharatviswa504) [inline phantomjsdriver.log]: I think a file is added by mistake.

**Comment by anuengineer** [comment]
+1, LGTM. Thanks for the patch, and thanks for the review. @bharatviswa504

**Pull request merged**